### ISSUE TITLE
Support for AMD and Intel devices

### ISSUE THREAD

**Issue by Syllo** [Syllo/nvtop#106]
I am opening this up because of [this thread of discussion on the intel-sgx mailing list](https://lists.freedesktop.org/archives/intel-gfx/2021-May/266057.html).

It seems that the GPU utilization for AMD and Intel will be available in future kernels, at least for the processes spawned by the user itself.

Marking this as a possible improvement once the support lands into the mainline kernel.

**Comment by thesunexpress** [comment]
Might be a better idea to just call it ' gputop ' instead then...

**Comment by el1-x** [comment]
@Syllo any update on this?

**Comment by Syllo** [comment]
I will keep this issue around as a meta-issue to centralize the information on other vendors / driver support.

**Comment by thomas-zimmerman** [comment]
I'm on a 5.17.4 kernel with a RX 6700 XT AMD GPU and the output that I get is:
```
No GPU to monitor.
```
This is on Ubuntu 22.04 with `nvtop_1.2.2-1_amd64.deb`. I built from source and I get the same output. Any thing I should run to get output on what's missing on this kernel to get output?

**Comment by Syllo** [comment]
@thomas-zimmerman I created #144 for your issue

**Comment by Syllo** [comment]
The Intel sysfs change has been proposed for Linux 5.19: https://lore.kernel.org/dri-devel/Ymkfy8FjsG2JrodK@tursulin-mobl2/

**Comment by servesh** [comment]
@Syllo Would there be any interest in adding support for Intel GPUs via level zero Sysman interface (https://spec.oneapi.io/level-zero/latest/sysman/PROG.html)?

**Comment by Syllo** [comment]
@servesh Thanks for the link. I was not aware that oneAPI provided such library.
This seems like a good opportunity to simplify the accelerator discovery and query.
I'll see if I can make this work.

**Comment by servesh** [comment]
@Syllo Sure, Iet me know once you have something that I can test and take a look. Happy to contribute to the code base as well.

**Comment by HTOgit** [comment]
@Syllo Is it possible to show RX/TX on AMD GPUs? I'm on Linux 5.19.

**Comment by Syllo** [comment]
@HTOgit
The support for the [pcie_bw file](https://dri.freedesktop.org/docs/drm/gpu/amdgpu.html#pcie-bw) is in place.
However this file is empty on the system I am testing (RX 6800XT, Kernel 5.19.6).

The only explanation that I see is that the support has not been implemented in the driver:
- The function that populates the `pcie_bw` file is [amdgpu_get_pcie_bw](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/pm/amdgpu_pm.c#L1558)
- The Navi implementation seems to be in [soc21.c](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc21.c) but the function `get_pcie_usage` that is required by `amdgpu_get_pcie_bw` is not registered in [soc21_asic_funcs](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc21.c#L512) contrary to [soc20_asic_funcs](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc15.c#L883) and [vega20_asic_funcs](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc15.c#L905).

I created a feature request for this [here](https://gitlab.freedesktop.org/drm/amd/-/issues/2232)

**Comment by poltimmer** [comment]
Are there any plans on adding apple silicon support?

**Comment by HTOgit** [comment]
> @HTOgit The support for the [pcie_bw file](https://dri.freedesktop.org/docs/drm/gpu/amdgpu.html#pcie-bw) is in place. However this file is empty on the system I am testing (RX 6800XT, Kernel 5.19.6).
> 
> The only explanation that I see is that the support has not been implemented in the driver:
> 
>     * The function that populates the `pcie_bw` file is [amdgpu_get_pcie_bw](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/pm/amdgpu_pm.c#L1558)
> 
>     * The Navi implementation seems to be in [soc21.c](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc21.c) but the function `get_pcie_usage` that is required by `amdgpu_get_pcie_bw` is not registered in [soc21_asic_funcs](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc21.c#L512) contrary to [soc20_asic_funcs](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc15.c#L883) and [vega20_asic_funcs](https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/amd/amdgpu/soc15.c#L905).
> 
> 
> I created a feature request for this [here](https://gitlab.freedesktop.org/drm/amd/-/issues/2232)
Sorry for not following up on the issue.

Are there any news to the topic? AMD announced to expose PCIe information to user space with Linux6.3 (https://www.phoronix.com/news/Linux-6.3-AMDGPU-New-uAPI). Would this help?

**Comment by qwertychouskie** [comment]
https://www.phoronix.com/news/Raspberry-Pi-GPU-Stats-Report

Would be nice to have V3D supported as well.

**Comment by sin-ack** [comment]
On kernel 6.7.4, when I run [`text-generation-webui`](https://github.com/oobabooga/text-generation-webui), the "TYPE" field only shows "Graphic" and the GPU usage is always 0%:

![image](https://github.com/Syllo/nvtop/assets/77421532/4f72bbf7-aa5c-4d05-a41e-5eb080695c92)

The same program on an Nvidia GPU shows up correctly as "Compute" and I can see the GPU usage fine.

**Comment by ProMix0** [comment]
> at least for the processes spawned by the user itself
How this limitation may be solved? Currently I had to run nvtop as root to see GPU loading by processes on me server. Is there a suitable capability, or at least can you add suid on installation (or section in README) to workaround this?

**Comment by mechovation** [comment]
AMD Radeon AI R9700, will show idle (0% GPU) if nvtop is started with no load.  

Once the a load/job is applied, %GPU pegs at 100%, even if idle power reduces, indicating that job is complete.
Job complete can also be confirmed in Ollama logs.

%GPU does not "release" (return to zero) until model time-out and it unloads from GPU memory
